managed/services/uievents/uievents.go

```diff
@@ -0,0 +1,292 @@
+		}
+		if p95 >= 10_000 {
+			dashboardStat.SlowDashboardsP95_10s = append(dashboardStat.SlowDashboardsP95_10s, stat.uid)
+		}
+	}
+
+	serializedDashboardStat, err := json.Marshal(dashboardStat)
+	if err != nil {
+		s.l.Error("failed to marshal to JSON")
+	}
+
+	return &pmmv1.ServerMetric_Metric{
+		Key:   "ui-events-dashboards",
+		Value: string(serializedDashboardStat),
+	}
+}
+
+func (s *Service) processComponentMetrics(componentsUsage map[string]*ComponentsUsageStat) *pmmv1.ServerMetric_Metric {
+	type Stat struct {
+		SlowComponentsP95_1s  []string `json:"slow_components_p95_1s"`
+		SlowComponentsP95_5s  []string `json:"slow_components_p95_5s"`
+		SlowComponentsP95_10s []string `json:"slow_components_p95_10s"`
+	}
+
+	if len(s.componentsUsage) > 0 {
+		return nil
+	}
+
+	componentsStat := &Stat{
+		SlowComponentsP95_1s:  []string{},
+		SlowComponentsP95_5s:  []string{},
+		SlowComponentsP95_10s: []string{},
+	}
+
+	// SlowComponentP95
+	keys := make([]string, 0, len(componentsUsage))
+	for key := range componentsUsage {
+		keys = append(keys, key)
+	}
+	sort.SliceStable(keys, func(i, j int) bool {
+		return componentsUsage[keys[i]].loadTime.ValueAtPercentile(95) > componentsUsage[keys[j]].loadTime.ValueAtPercentile(95)
+	})
+	for i := 0; i < len(keys); i++ {
+		sortedKey := keys[i]
+		stat := componentsUsage[sortedKey]
+		p95 := stat.loadTime.ValueAtPercentile(95)
+		if p95 >= 1_000 {
+			componentsStat.SlowComponentsP95_1s = append(componentsStat.SlowComponentsP95_1s, stat.uid)
+		}
+		if p95 >= 5_000 {
+			componentsStat.SlowComponentsP95_5s = append(componentsStat.SlowComponentsP95_5s, stat.uid)
+		}
+		if p95 >= 10_000 {
+			componentsStat.SlowComponentsP95_10s = append(componentsStat.SlowComponentsP95_10s, stat.uid)
+		}
+	}
+
+	serializedComponentsStat, err := json.Marshal(componentsStat)
+	if err != nil {
+		s.l.Error("failed to marshal to JSON")
+	}
+	return &pmmv1.ServerMetric_Metric{
+		Key:   "ui-events-components",
+		Value: string(serializedComponentsStat),
+	}
+}
+
+func (s *Service) processUserFlowEvents(events []*uievents.UserFlowEvent) []*pmmv1.ServerMetric_Metric {
+	var result []*pmmv1.ServerMetric_Metric
+	for _, event := range events {
+		marshal, err := json.Marshal(event)
+		if err != nil {
+			s.l.Error("failed to marshal to JSON")
+		}
+		result = append(result, &pmmv1.ServerMetric_Metric{
+			Key:   "ui-events-user-flow",
+			Value: string(marshal),
+		})
+	}
+	return result
+}
+
+// Store stores metrics for further processing and sending to Portal.
+func (s *Service) Store(ctx context.Context, request *uievents.StoreRequest) (*uievents.StoreResponse, error) {
+	for _, event := range request.DashboardUsage {
+		stat, ok := s.dashboardUsage[event.Uid]
+		if !ok {
+			stat = &DashboardUsageStat{
+				title:    event.Title,
+				uid:      event.Uid,
+				useCount: 0,
+				loadTime: hdrhistogram.New(lowersValue, highestValue, numberOfSignificantValueDigits),
+			}
+			s.dashboardUsage[event.Uid] = stat
```

Comment:
Not protected by mutex

Reply:
there is not race in store, but I'll add

Reply:
This test was failing with datarace before latest changes:
```
func Test(t *testing.T) {
	t.Parallel()

	s, err := New()
	require.NoError(t, err)

	t.Run("A", func(t *testing.T) {
		t.Parallel()
		_, err := s.Store(context.Background(), &uievents.StoreRequest{
			Notifications: []*uievents.NotificationEvent{
				{
					Title:          "test",
					Text:           "test",
					Location:       "test",
					LocationParams: "test",
				},
			},
			DashboardUsage: []*uievents.DashboardUsageEvent{
				{
					Uid:            "test",
					Title:          "test",
					Tags:           nil,
					LoadTime:       0,
					Location:       "test",
					LocationParams: "test",
				},
			},
		})
		require.NoError(t, err)
	})

	t.Run("B", func(t *testing.T) {
		t.Parallel()
		_, err := s.FetchMetrics(context.Background(), telemetry.Config{})
		require.NoError(t, err)
	})
}
```